graph/src/util/futures.rs

```diff
@@ -1,74 +1,288 @@
 use slog::Logger;
+use std::fmt::Debug;
+use std::marker::PhantomData;
+use std::sync::Arc;
 use std::time::{Duration, Instant};
 use tokio::prelude::*;
 use tokio::timer::DeadlineError;
 use tokio_retry::strategy::{jitter, ExponentialBackoff};
 use tokio_retry::Error as RetryError;
 use tokio_retry::Retry;
 
-/// Repeatedly invoke `try_it` until the future it returns resolves to an `Ok`.
-/// `operation_name` is used in log messages.
-/// Warnings will be logged after failed attempts.
-pub fn with_retry<F, R, I, E>(
-    logger: Logger,
+/// Generic helper function for retrying async operations with built-in logging.
+///
+/// To use this helper, do the following:
+///
+/// 1. Call this function with an operation name (used for logging) and a `Logger`.
+/// 2. Chain a call to `.when_err()` or `.when(...)`.
```

Comment:
Oh yeah, I think that's a good idea. I made limit and timeout required since it's something I want to force myself to think through every time, but the retry condition is something you tend to think through automatically

Reply:
Done